utils/cloudinfo/cloudinfo.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudinfo
+
+import (
+	"net/http"
+)
+
+type CloudProvider int
+
+const (
+	GCE CloudProvider = iota
+	AWS
+	BAREMETAL
+	UNKNOWN
+)
+
+const NOINSTANCE = "NA"
+
+type CloudInfo interface {
+	GetCloudProvider() (CloudProvider, error)
+	GetInstanceType(CloudProvider) (string, error)
+}
+
+type realCloudInfo struct{}
+
+func NewRealCloudInfo() (CloudInfo, error) {
+	return &realCloudInfo{}, nil
+}
+
+func (self *realCloudInfo) GetCloudProvider() (CloudProvider, error) {
+	switch {
+	case inGCE():
+		return GCE, nil
+	case inAWS():
+		return AWS, nil
+	case inBareMetal():
+		return BAREMETAL, nil
+	}
+	return UNKNOWN, nil
+}
+
+func (self *realCloudInfo) GetInstanceType(cp CloudProvider) (string, error) {
```

Comment:
If we save cloudprovider and instance type internally, we don't need to accept cloudprovider as a parameter.